Email templates customizable

drever opened the pull request:
Hello Christian,

please have a first look at the code! I moved the email templates to a separate class. I did not get a setup running on my local machine. The configuration.nix is a nix expression which starts a server. Unfortunately the apache server crashed with a segfault. Anyways, the tests are working, I'm curious if the email templates also work!

Cheers,
Johannes

cgogolin commented:
Excellent! I will give it a look on the weekend or next week.

cgogolin commented:
Cool! This is looking very good! I especially like the elegant way to render the available shortcodes.

I just have a few very minor comments:

1. You chose the format `#SHORCODE#` for the shortcodes. I think I can see why you did that (because square brackets need to be escaped in regular expressions), but WP users are typically used to shortcodes with the `[shortcode]` format (see [here](https://en.support.wordpress.com/shortcodes/) for example). I think it would be good to follow that convention, even though our "shortcodes" are not really shortcodes, they only behave like ones in the limited context of the email templates. Do we actually need regular expression search and replace? Wouldn't a plain vanilla `str_replace` be sufficient?

2. In line 127 and 140 of `class-o3po-settings.php` you have
```
"In case you have an ORCID you can go to http://search.crossref.org/?q=#ORCID# to conveniently add your new publication to your profile.\n\n"
```
but the shortcode here should be #DOI# instead of #ORCID#. And maybe there is a comma missing after `Dear #AUTHORS#`. 

3. I quite like that PHP has not problems with "dangling" `,`s in definitions of lists/dictionaries and thus prefer them to be at the end of each line. Sorry for being pedantic, but if you don't mind, it would be nice if we could keep the code consistent in this respect. 

I have no experience with nix expressions. Does it really make sense to include the nix into the repository? Especially if it produces a segfault? In the long run, you will need a test system. I recommend using the `onepress` theme, that is the one we are using at Quantum and O-3PO is not yet fully theme independent (anyway, this shouldn't explain the segfault...). Maybe there is an issue with the PHP version?

cgogolin commented:
Nix looks pretty cool! Would be nice to get that to work.

cgogolin commented:
A last comment:

4.  Some journals will not have a Executive Board, but instead an Editor in Chief. Maybe we can have an alias shortcode for that or call the whole things "publisher responsible" or something like that? What do you think?

drever commented:
Thanks for comments! I will answer them below..

drever commented:
> Does it really make sense to include the nix into the repository? 

I'm working with nixos, so it would be really convenient for me to have nix support. And I think nix is pretty cool :). I moved this to a separate issue #30. Let's see how it works out, but I don't want to spend too much time on it. I could also use MacOS, what would you suggest?

drever commented:
> 2. [...] but the shortcode here should be #DOI# instead of #ORCID#. And maybe there is a comma missing after Dear #AUTHORS#.

Changed it to   

         DOI: [doi_url_prefix][doi]
         In case you have an ORCID you can go to http://search.crossref.org/?q=[doi]

Please have a look if this is correct, the old implementation replaces the `/` with `2#F`. The URLs in the current test code do work though.

drever commented:
> 3. I quite like that PHP has not problems with "dangling" ,s in definitions of lists/dictionaries and thus prefer them to be at the end of each line. Sorry for being pedantic, but if you don't mind, it would be nice if we could keep the code consistent in this respect.

The leading comma style is often encountered in the Haskell community ([e.g. here](https://www.joachim-breitner.de/blog/739-Avoid_the_dilemma_of_the_trailing_comma)). I prefer that style but changed it for the sake of consistency. Maybe we should use a  code formatter to avoid bike shedding.

drever commented:
> Some journals will not have a Executive Board, but instead an Editor in Chief. Maybe we can have an alias shortcode for that or call the whole things "publisher responsible" or something like that? What do you think?

The "executive board" is hard coded anyways right now. I would suggest to have a short code [executive_board] and [editor_in_chief] and then they can be referred to freely.

drever commented:
I'm done for now. Let me know if I forgot to address something are something else came up. Cheers!

drever commented:
> 3. [..] How do we deal with that? Shall we offer an additional short code [doi-hex-encoded]?

> 5. +1 for having both [executive_board] and [editor_in_chief]. This is not yet implemented, right? Then I wait with merging until that is done.

I added the new short code. This should work for now. From where does the information come from? I used `get_journal_property('executive_board')`, but I'm not sure if this is correct. I broke the tests, do I have to add it to `required_journal_config_properties`?

drever commented on tests/o3po-email-templates-test.php:
```diff
@@ -16,14 +16,14 @@ public function test_self_notification_subject(){
-                          "DOI:     test-doi\n".
-                          "NOT IN TEMPLATE! test-doi-hex-encoded",
+                          "DOI:     test-doi/with-special-char\n".
+                          "NOT IN TEMPLATE! test-doi%2Fwith-special-char",
```
Nice catch!

drever commented:
The issue looks good for to me for now. We can also have a chat on how to proceed if you like!

cgogolin commented:
The settings were not actually saved. I fixed that and added a test that should make it harder to overlook adding settings to the `all_settings_fields_map` in the future. For security, only the settings in there, and with a a specified sanitation function are actually saved.

cgogolin commented:
The last few things before we can merge this:

1. How do we want to sanitize the user input in the template settings? 

2. There should be a separate `'author_notification_secondary_subject_template'`. I will add a test that enforces this.

3. While we are at it, let's also fix the restrictions on the DOI prefix discussed in #31. I think the primary and secondary DOI prefix should only be allowed to contain alphanumeric characters and `-`. Any thoughts on this?

cgogolin commented:
4. Please add docstrings to all the functions in the new class. For `@since` please use `2.2.2+` I will replace that with the version of the next release (probably `0.3.0`) when releasing.

cgogolin commented:
Thanks Johannes! I consider 2 done.

I have already implemented the restrictions on the DOI prefix:
```
    private function sanitize_doi_prefix( $doi_prefix ) {

        $doi_prefix = trim($doi_prefix);
        if(preg_match('/^[-0-9]*$/', $doi_prefix))
            return $doi_prefix;
        else
            "";
    }
```
I think I will open a new issue on better user feedback for illegal entries in settings fields and sanitization, so for the purpose of this issue I will consider also 1 and 3 fixed, unless you object.

When you try to build the docs `make docs`, you will see that there remain the following errors: 
```
Parsing /home/cgogolin/src/o3po/o3po/includes/class-o3po-email-templates.php
  Argument $journal is missing from the Docblock of self_notification_subject
  Name of argument $journal does not match with the DocBlock's name $template in self_notification_subject()
  Argument $publication_type_name is missing from the Docblock of self_notification_subject
  Name of argument $publication_type_name does not match with the DocBlock's name $template in self_notification_subject()
  Name of argument $email_template does not match with the DocBlock's name $template in self_notification_body()
  Argument $journal is missing from the Docblock of self_notification_body
  Name of argument $journal does not match with the DocBlock's name $template in self_notification_body()
  Argument $publication_type_name is missing from the Docblock of self_notification_body
  Name of argument $publication_type_name does not match with the DocBlock's name $template in self_notification_body()
  Argument $title is missing from the Docblock of self_notification_body
  Name of argument $title does not match with the DocBlock's name $template in self_notification_body()
  Argument $authors is missing from the Docblock of self_notification_body
  Name of argument $authors does not match with the DocBlock's name $template in self_notification_body()
  Argument $url is missing from the Docblock of self_notification_body
  Name of argument $url does not match with the DocBlock's name $template in self_notification_body()
  Argument $doi_url_prefix is missing from the Docblock of self_notification_body
  Name of argument $doi_url_prefix does not match with the DocBlock's name $template in self_notification_body()
  Argument $doi is missing from the Docblock of self_notification_body
  Name of argument $doi does not match with the DocBlock's name $template in self_notification_body()
  Argument $journal is missing from the Docblock of author_notification_subject
  Name of argument $journal does not match with the DocBlock's name $template in author_notification_subject()
  Argument $publication_type_name is missing from the Docblock of author_notification_subject
  Name of argument $publication_type_name does not match with the DocBlock's name $template in author_notification_subject()
  Name of argument $email_template does not match with the DocBlock's name $template in author_notification_body()
  Argument $journal is missing from the Docblock of author_notification_body
  Name of argument $journal does not match with the DocBlock's name $template in author_notification_body()
  Argument $executive_board is missing from the Docblock of author_notification_body
  Name of argument $executive_board does not match with the DocBlock's name $template in author_notification_body()
  Argument $editor_in_chief is missing from the Docblock of author_notification_body
  Name of argument $editor_in_chief does not match with the DocBlock's name $template in author_notification_body()
  Argument $publisher_email is missing from the Docblock of author_notification_body
  Name of argument $publisher_email does not match with the DocBlock's name $template in author_notification_body()
  Argument $publication_type_name is missing from the Docblock of author_notification_body
  Name of argument $publication_type_name does not match with the DocBlock's name $template in author_notification_body()
  Argument $title is missing from the Docblock of author_notification_body
  Name of argument $title does not match with the DocBlock's name $template in author_notification_body()
  Argument $authors is missing from the Docblock of author_notification_body
  Name of argument $authors does not match with the DocBlock's name $template in author_notification_body()
  Argument $url is missing from the Docblock of author_notification_body
  Name of argument $url does not match with the DocBlock's name $template in author_notification_body()
  Argument $doi_url_prefix is missing from the Docblock of author_notification_body
  Name of argument $doi_url_prefix does not match with the DocBlock's name $template in author_notification_body()
  Argument $doi is missing from the Docblock of author_notification_body
  Name of argument $doi does not match with the DocBlock's name $template in author_notification_body()
  Argument $journal_reference is missing from the Docblock of author_notification_body
  Name of argument $journal_reference does not match with the DocBlock's name $template in author_notification_body()
  Argument $journal is missing from the Docblock of fermats_library_notification_subject
  Name of argument $journal does not match with the DocBlock's name $template in fermats_library_notification_subject()
  Argument $publication_type_name is missing from the Docblock of fermats_library_notification_subject
  Name of argument $publication_type_name does not match with the DocBlock's name $template in fermats_library_notification_subject()
  Name of argument $email_template does not match with the DocBlock's name $template in fermats_library_notification_body()
  Argument $journal is missing from the Docblock of fermats_library_notification_body
  Name of argument $journal does not match with the DocBlock's name $template in fermats_library_notification_body()
  Argument $publication_type_name is missing from the Docblock of fermats_library_notification_body
  Name of argument $publication_type_name does not match with the DocBlock's name $template in fermats_library_notification_body()
  Argument $title is missing from the Docblock of fermats_library_notification_body
  Name of argument $title does not match with the DocBlock's name $template in fermats_library_notification_body()
  Argument $authors is missing from the Docblock of fermats_library_notification_body
  Name of argument $authors does not match with the DocBlock's name $template in fermats_library_notification_body()
  Argument $url is missing from the Docblock of fermats_library_notification_body
  Name of argument $url does not match with the DocBlock's name $template in fermats_library_notification_body()
  Argument $doi is missing from the Docblock of fermats_library_notification_body
  Name of argument $doi does not match with the DocBlock's name $template in fermats_library_notification_body()
  Argument $fermats_library_permalink is missing from the Docblock of fermats_library_notification_body
  Name of argument $fermats_library_permalink does not match with the DocBlock's name $template in fermats_library_notification_body()
```

drever commented:
Hi Christian!

I fixed the documentation. I also added some `[doi_url_prefix]` to the templates. I'm not sure what we agreed on so please have a look.

cgogolin commented:
I added are review comment, please have a look at that. 

Could you please have a look an test the new settings in your test environment? Are the settings saved? Does the settings page look reasonable? Do you actually receive the emails? Do they look correct?

cgogolin commented:
I still get an error when making the docs. Please try running `make docs`.

drever commented:
> I added are review comment, please have a look at that.

I don't see a reviewer comment.

> Could you please have a look an test the new settings in your test environment? Are the settings saved? Does the settings page look reasonable? Do you actually receive the emails? Do they look correct?

My test environment is still brittle (#30), that's why I rely on the integration tests. I guess I should move this up higher on my priority list.

cgogolin commented:
Sorry, I forgot to submit the review. Not you should be able to see it. You really need a working testing environment. Otherwise this will get very inefficient...
